refactor AU units

psychocoderHPC opened the pull request:
- add conversion au units -> si
- add conversion au unis -> pic

**Please review very carefully!!!!!!**

# Review

Please check the first commit only. The second commit is changing the access `.conv.*` to `.conv().*`. This is required because nvcc is not able to pull the constexpr members correctly into the device code.

psychocoderHPC committed:
refactor AU units

- add conversion au units -> si
- add conversion au unis -> pic

ci: picongpu

psychocoderHPC committed:
change interface `conv.*` to `conv().*

nvcc is not possible to pull static constexpr member into the device
code, therefore acessing a function is required.

BrianMarre commented:
Still in the process of the reviewing the PR, but there is one thing I am already really unhappy with.
This PR only provides conversion methods for Atomic Units instead of treating AtomicUnits as a separate equally valid third unit system.

This in my eyes the wrong approach and leads to combining orthogonal sets of functions elements into one class and a large number of conversion methods.

The root problem seems to be that the PR tries to press both code internal units and AtomicUnits into one class, although AtomicUnits are logically independent of the code internal units and should therefore be represented as a separate class with it's own conversion functions, instead of mashing them into one general set of conversion functions from SI to whatever.

My suggestion would be to define for internalPicUnits and atomicUnits each a class holding only the 4 base quantities and their SI-values.
Namely for internalPicUnits
- massElectron
- speedOfLight
- simulationTimeStepLength
- electronCharge

And for atomicUnits:
- atomicUnitEFieldStrength
- RydbergEnergy
- AtomicUnitTime
- (massElectron)

and then based on these define for each an daughter class that defines all the derived units in their respective unit systems, e.g. internalPicUnitEField, ... .

In addition it would be nice to define unit prefixes in one central location to provide generic conversion methods to avoid the magic prefix conversion numbers.

BrianMarre commented on include/picongpu/particles/atomicPhysics/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -248,7 +248,7 @@ namespace picongpu::particles::atomicPhysics::initElectrons
             // kg * m^2/s^2 * keV/J * 1e3 = J/J * eV = eV
             float_64 const mc2_Electron = static_cast<float_64>(massElectron) * sim.unit.mass()
-                * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u) * sim.si.conv.joule2ev(1.0);
+                * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u) * sim.si.conv().joule2ev(1.0);
```
same as above

BrianMarre commented on include/picongpu/particles/atomicPhysics/kernel/DecelerateElectrons.kernel:
```diff
@@ -134,15 +134,15 @@ namespace picongpu::particles::atomicPhysics::kernel
                     float_X const newElectronMomentum
                         = math::sqrt(newEnergyElectron * (newEnergyElectron + 2 * mcSquaredElectron)) * scalingFactor;
-                    // AU = ATOMIC_UNIT_ENERGY
+                    // AU = sim.si.conv().auEnergy2Joule()
```
here the meaning of the comment is lost since instead of indicating a used unit, with physical meaning, it now indicates some magic scaling factor that makes understanding the comment harder.

BrianMarre commented on include/picongpu/particles/atomicPhysics/rateCalculation/BoundFreeFieldTransitionRates.hpp:
```diff
@@ -92,8 +92,8 @@ namespace picongpu::particles::atomicPhysics::rateCalculation
             float_X const effectivePrincipalQuantumNumber
-                = screenedCharge / math::sqrt(2._X * ionizationEnergy * picongpu::UNITCONV_eV_to_AU);
-            float_X const eFieldNorm_AU = eFieldNorm / ATOMIC_UNIT_EFIELD;
+                = screenedCharge / math::sqrt(2._X * ionizationEnergy * sim.si.conv().ev2auEnergy(1.0));
```
```suggestion
                = screenedCharge / math::sqrt(2._X * sim.si.conv().ev2auEnergy(ionizationEnergy));
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/rateCalculation/BoundFreeFieldTransitionRates.hpp:
```diff
@@ -112,7 +112,7 @@ namespace picongpu::particles::atomicPhysics::rateCalculation
                 rateADK_AU *= math::sqrt(3._X * nEffCubed * eFieldNorm_AU / (pi * screenedChargeCubed));
 
-            return rateADK_AU / ATOMIC_UNIT_TIME;
+            return rateADK_AU / sim.pic.conv().auTime2Time(1.0);
```
this line demonstrated nicely why defining an AtomicUnitTime explicitly is to be preferred.

psychocoderHPC commented:
> Still in the process of the reviewing the PR, but there is one thing I am already really unhappy with. This PR only provides conversion methods for Atomic Units instead of treating AtomicUnits as a separate equally valid third unit system.

I agree with you that the current approach is not very nice. It is fully orientated on the current approach.
Current normalizations are from a time long before constexpr and other helpful features.
This PR is not naming to solve the unit system issue, all the current refactorings have the goal of removing cyclic dependencies or simply removing the `*.unitless` files.

I opened the issue https://github.com/ComputationalRadiationPhysics/picongpu/issues/5135 for the unit system, but to be fair nobody was interested in this topic until now and without contribution from others I do not see that this will be solved in the next years.

BrianMarre commented on include/picongpu/particles/atomicPhysics/ionizationPotentialDepression/StewartPyattIPD.hpp:
```diff
@@ -211,7 +211,7 @@ namespace picongpu::particles::atomicPhysics::ionizationPotentialDepression
             constexpr float_X eV = static_cast<float_X>(
                 picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u)
-                / pmacc::math::cPow(picongpu::sim.unit.time(), 2u) * sim.si.conv.joule2ev(1.0));
+                / pmacc::math::cPow(picongpu::sim.unit.time(), 2u) * sim.si.conv().joule2ev(1.0));
```
Misuse of a conversion function, what we actually want here the value of 1eV in internal PIC units, I would suggest to define something along the line
```c++
struct SIUnits
{
    ...
   //! get 1eV in J
   constexpr float_64 geteV()
   {
        return 1.602176634e-19;
    }
```
and replace the lines 212 -214 with the following.

```suggestion
     constexpr float_X eV = static_cast<float_X>(picongpu::sim.si.geteV() / piconggpu::sim.unit.energy());
```
or write a dedicated conversion method that converts internal energy to eV

BrianMarre commented on include/picongpu/param/ionizer.param:
```diff
@@ -493,7 +493,7 @@ namespace picongpu
                 constexpr float_X CUTOFF_MAX_ENERGY_KEV = 50.0;
                 /** cutoff energy for electron "temperature" calculation in SI units*/
-                constexpr float_X CUTOFF_MAX_ENERGY = CUTOFF_MAX_ENERGY_KEV * sim.si.conv.ev2Joule(1.0e3);
+                constexpr float_X CUTOFF_MAX_ENERGY = CUTOFF_MAX_ENERGY_KEV * sim.si.conv().ev2Joule(1.0e3);
```
```suggestion
                constexpr float_X CUTOFF_MAX_ENERGY = sim.si.conv().ev2Joule(CUTOFF_MAX_ENERGY_KEV * 1.0e3);
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/GetPhysicalEnergy.hpp:
```diff
@@ -69,7 +69,7 @@ namespace picongpu::particles::atomicPhysics
             constexpr float_X eV = static_cast<float_X>(
                 picongpu::sim.unit.mass() * picongpu::sim.si.getSpeedOfLight() * picongpu::sim.si.getSpeedOfLight()
-                * sim.si.conv.joule2ev(1.0));
+                * sim.si.conv().joule2ev(1.0));
```
```suggestion
    constexpr float_X eV = static_cast<float_X>(sim.si.convert().joule2eV(picongpu::sim.unit.mass() * picongpu::sim.si.getSpeedOfLight() * picongpu::sim.si.getSpeedOfLight()));
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/initElectrons/Inelastic2BodyCollisionFromCoMoving.hpp:
```diff
@@ -199,7 +199,7 @@ namespace picongpu::particles::atomicPhysics::initElectrons
             // kg * m^2/s^2 * keV/J * 1e3 = J/J * eV = eV
             float_64 const mc2_Ion = static_cast<float_64>(massIon) * sim.unit.mass()
-                * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u) * sim.si.conv.joule2ev(1.0);
+                * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u) * sim.si.conv().joule2ev(1.0);
```
```suggestion
                float_64 const mc2_Ion = static_cast<float_64>(massIon) * sim.si.conv().joule2ev(sim.unit.mass()
                * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u));
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/ionizationPotentialDepression/kernel/CalculateIPDInput.kernel:
```diff
@@ -152,7 +152,7 @@ namespace picongpu::particles::atomicPhysics::ionizationPotentialDepression::ker
                     constexpr float_X eV = static_cast<float_X>(
                         picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u)
-                        / pmacc::math::cPow(picongpu::sim.unit.time(), 2u) * sim.si.conv.joule2ev(1.0));
+                        / pmacc::math::cPow(picongpu::sim.unit.time(), 2u) * sim.si.conv().joule2ev(1.0));
```
```suggestion
    constexpr float_X eV = static_cast<float_X>(sim.si.conv().joule2eV(picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u)));
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/kernel/DecelerateElectrons.kernel:
```diff
@@ -115,7 +115,7 @@ namespace picongpu::particles::atomicPhysics::kernel
                         static_cast<float_64>(picongpu::traits::frame::getMass<typename T_Electron::FrameType>())
                         * sim.unit.mass() * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u)
-                        * sim.si.conv.joule2ev(1.0));
+                        * sim.si.conv().joule2ev(1.0));
```
```suggestion
* sim.si.conv().joule2ev(sim.unit.mass() * pmacc::math::cPow(picongpu::sim.si.getSpeedOfLight(), 2u));
```

BrianMarre commented on include/picongpu/particles/manipulators/unary/Temperature.hpp:
```diff
@@ -71,7 +71,7 @@ namespace picongpu
                                  */
                                 float_X const energy
-                                    = (temperatureKeV * sim.si.conv.ev2Joule(1.0e3)) / sim.unit.energy();
+                                    = (temperatureKeV * sim.si.conv().ev2Joule(1.0e3)) / sim.unit.energy();
```
```suggestion
                                    = sim.si.conv().ev2Joule(temperatureKeV * 1.0e3) / sim.unit.energy();
```

BrianMarre commented on include/picongpu/plugins/BinEnergyParticles.x.cpp:
```diff
@@ -427,8 +427,8 @@ namespace picongpu
-            float_X const minEnergy = minEnergy_keV * sim.si.conv.ev2Joule(1.0e3) / sim.unit.energy();
-            float_X const maxEnergy = maxEnergy_keV * sim.si.conv.ev2Joule(1.0e3) / sim.unit.energy();
+            float_X const minEnergy = minEnergy_keV * sim.si.conv().ev2Joule(1.0e3) / sim.unit.energy();
+            float_X const maxEnergy = maxEnergy_keV * sim.si.conv().ev2Joule(1.0e3) / sim.unit.energy();
```
```suggestion
            float_X const keV = static_cast<float_X>(sim.si.conv().ev2Joule(1.0e3) / sim.unit.energy());
            float_X const minEnergy = minEnergy_keV * keV;
            float_X const maxEnergy = maxEnergy_keV * keV;
```

BrianMarre commented on include/picongpu/plugins/particleCalorimeter/ParticleCalorimeter.x.cpp:
```diff
@@ -119,8 +119,8 @@ namespace picongpu
-                const float_64 minEnergy_keV = minEnergy_SI * sim.si.conv.joule2ev(1.0e-3);
-                const float_64 maxEnergy_keV = maxEnergy_SI * sim.si.conv.joule2ev(1.0e-3);
+                const float_64 minEnergy_keV = minEnergy_SI * sim.si.conv().joule2ev(1.0e-3);
+                const float_64 maxEnergy_keV = maxEnergy_SI * sim.si.conv().joule2ev(1.0e-3);
```
```suggestion
                float64 const keVperJ = sim.si.conv().joule2eV(1.0e-3);

                float_64 const minEnergy_keV = minEnergy_SI * keVperJ;
                float_64 const maxEnergy_keV = maxEnergy_SI * keVperJ;
```
still not really happy with but don't see a better version

BrianMarre commented on include/picongpu/plugins/particleCalorimeter/ParticleCalorimeter.x.cpp:
```diff
@@ -301,8 +301,8 @@ namespace picongpu
-            const float_64 minEnergy_SI = this->minEnergy * sim.si.conv.ev2Joule(1.0e3);
-            const float_64 maxEnergy_SI = this->maxEnergy * sim.si.conv.ev2Joule(1.0e3);
+            const float_64 minEnergy_SI = this->minEnergy * sim.si.conv().ev2Joule(1.0e3);
+            const float_64 maxEnergy_SI = this->maxEnergy * sim.si.conv().ev2Joule(1.0e3);
```
```suggestion
            float_64 const keV = sim.si.conv().ev2Joule(1.0e3);
            
            float_64 const minEnergy_SI = this->minEnergy * keV;
            float_64 const maxEnergy_SI = this->maxEnergy * keV;
```

BrianMarre commented on share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:
```diff
@@ -82,7 +82,7 @@ namespace picongpu
                  *  unit: keV
                  */
                 static constexpr float_64 temperature = sim.si.getElectronMass() * sim.si.getSpeedOfLight()
-                    * sim.si.getSpeedOfLight() * sim.si.conv.joule2ev(1.0e-3) * 0.0000002;
+                    * sim.si.getSpeedOfLight() * sim.si.conv().joule2ev(1.0e-3) * 0.0000002;
```
```suggestion
                static constexpr float_64 temperature = sim.si.conv().joule2ev(sim.si.getElectronMass() * sim.si.getSpeedOfLight() * sim.si.getSpeedOfLight()) * 0.0002;
```

BrianMarre commented on share/picongpu/tests/CollisionsBeamRelaxation/include/picongpu/param/particle.param:
```diff
@@ -96,7 +96,7 @@ namespace picongpu
                  *  unit: keV
                  */
                 static constexpr float_64 temperature = sim.si.getElectronMass() * sim.si.getSpeedOfLight()
-                    * sim.si.getSpeedOfLight() * sim.si.conv.joule2ev(1.0e-3) * 0.00002;
+                    * sim.si.getSpeedOfLight() * sim.si.conv().joule2ev(1.0e-3) * 0.00002;
```
```suggestion
                static constexpr float_64 temperature = sim.si.conv().joule2ev(sim.si.getElectronMass() * sim.si.getSpeedOfLight()
                    * sim.si.getSpeedOfLight()) * 0.02;
```

BrianMarre commented on share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param:
```diff
@@ -69,7 +69,7 @@ namespace picongpu
                  *  unit: keV
                  */
                 static constexpr float_64 temperature = sim.si.getElectronMass() * sim.si.getSpeedOfLight()
-                    * sim.si.getSpeedOfLight() * sim.si.conv.joule2ev(1.0e-3) * 0.0002;
+                    * sim.si.getSpeedOfLight() * sim.si.conv().joule2ev(1.0e-3) * 0.0002;
```
```suggestion
                static constexpr float_64 temperature = sim.si.conv().joule2ev(sim.si.getElectronMass() * sim.si.getSpeedOfLight() * sim.si.getSpeedOfLight()) * 0.2;
```

BrianMarre commented on share/picongpu/tests/CollisionsThermalisation/include/picongpu/param/particle.param:
```diff
@@ -83,7 +83,7 @@ namespace picongpu
                  *  unit: keV
                  */
                 static constexpr float_64 temperature = sim.si.getElectronMass() * sim.si.getSpeedOfLight()
-                    * sim.si.getSpeedOfLight() * sim.si.conv.joule2ev(1.0e-3) * 0.00018;
+                    * sim.si.getSpeedOfLight() * sim.si.conv().joule2ev(1.0e-3) * 0.00018;
```
```suggestion
                static constexpr float_64 temperature = sim.si.conv().joule2ev(sim.si.getElectronMass() * sim.si.getSpeedOfLight() * sim.si.getSpeedOfLight()) * 0.18;
```

BrianMarre commented on share/picongpu/tests/compileAtomicPhysics/include/picongpu/param/ionizer.param:
```diff
@@ -193,7 +193,7 @@ namespace picongpu
                 constexpr float_X CUTOFF_MAX_ENERGY_KEV = 50.0;
                 /** cutoff energy for electron "temperature" calculation in SI units*/
-                constexpr float_X CUTOFF_MAX_ENERGY = CUTOFF_MAX_ENERGY_KEV * sim.si.conv.ev2Joule(1.0e3);
+                constexpr float_X CUTOFF_MAX_ENERGY = CUTOFF_MAX_ENERGY_KEV * sim.si.conv().ev2Joule(1.0e3);
```
```suggestion
                constexpr float_X CUTOFF_MAX_ENERGY = sim.si.conv().ev2Joule(CUTOFF_MAX_ENERGY_KEV * 1.0e3);
```

BrianMarre reviewed (approved):
Found No bugs, seems to be function wise correct.

psychocoderHPC committed:
apply review comments

- add `automicUnit`
- rename `ev` to `eV`

BrianMarre commented on include/picongpu/particles/atomicPhysics/ionizationPotentialDepression/kernel/CalculateIPDInput.kernel:
```diff
@@ -150,9 +150,8 @@ namespace picongpu::particles::atomicPhysics::ionizationPotentialDepression::ker
-                    constexpr float_X eV = static_cast<float_X>(
-                        picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u)
-                        / pmacc::math::cPow(picongpu::sim.unit.time(), 2u) * sim.si.conv.joule2ev(1.0));
+                    constexpr float_X eV = static_cast<float_X>(sim.si.conv().joule2eV(
+                        picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u)));
```
that's a bug, it should read
```
constexpr float_X eV = static_cast<float_X>(sim.si.conv().joule2eV(
    picongpu::sim.unit.mass() * pmacc::math::cPow(picongpu::sim.unit.length(), 2u) / pmacc::math::cPow(picongpu::sim.unit.time(), 2u)));
```
or even shorter
```suggestion
                    constexpr float_X eV = static_cast<float_X>(picongpu::sim.si.conv().joule2eV(picongpu::sim.unit.energy()));
```

psychocoderHPC replied:
Should it maybe `sim.pic.get_eV();` because I assume you would like to have a picongpu eV?

BrianMarre replied:
Yes that is even better!

BrianMarre commented on include/picongpu/particles/atomicPhysics/kernel/DecelerateElectrons.kernel:
```diff
@@ -134,15 +134,15 @@ namespace picongpu::particles::atomicPhysics::kernel
 
                     constexpr float_X conversionEnergy
-                        = static_cast<float_X>(sim.si.conv.ev2Joule(1.0) / sim.unit.energy());
+                        = static_cast<float_X>(sim.si.conv().eV2Joule(1.0) / sim.unit.energy());
```
```suggestion
                        = static_cast<float_X>(picongpu::sim.pic.get_eV());
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/rateCalculation/BoundBoundTransitionRates.hpp:
```diff
@@ -390,7 +390,7 @@ namespace picongpu::particles::atomicPhysics::rateCalculation
             // J/(eV) / (Js) * s/sim.unit.time() = J/J * s/s * 1/(eV * sim.unit.time())
             constexpr float_X scalingConstantPhotonFrequency
-                = static_cast<float_X>(sim.si.conv.ev2Joule(1.0) / (2 * pi * hbar_SI) * picongpu::sim.unit.time());
+                = static_cast<float_X>(sim.si.conv().eV2Joule(1.0) / (2 * pi * hbar_SI) * picongpu::sim.unit.time());
```
```suggestion
                = static_cast<float_X>(picongpu::sim.si.get_eV() / (2 * pi * hbar_SI) * picongpu::sim.unit.time());
```

BrianMarre commented on include/picongpu/particles/atomicPhysics/rateCalculation/CollisionalRate.hpp:
```diff
@@ -46,7 +46,7 @@ namespace picongpu::particles2::atomicPhysics::rateCalculation
         constexpr float_64 m_e_SI = picongpu::sim.si.getElectronMass(); // [kg]
 
-        constexpr float_64 electronRestMassEnergy = m_e_SI * c_SI * c_SI / sim.si.conv.ev2Joule(1.0);
+        constexpr float_64 electronRestMassEnergy = m_e_SI * c_SI * c_SI / sim.si.conv().eV2Joule(1.0);
```
```suggestion
        constexpr float_64 electronRestMassEnergy = m_e_SI * c_SI * c_SI / picongpu::sim.si.get_eV();
```

BrianMarre commented on include/picongpu/unitless/simulation.unitless:
```diff
@@ -127,6 +135,50 @@ namespace picongpu
+                 * @{
+                 */
+                template<typename T_Type = float_X>
+                constexpr T_Type eV2Joule(T_Type eV = 1.0) const
```
```suggestion
                constexpr T_Type eV2picEnergy(T_Type eV = 1.0) const
```

BrianMarre replied:
but be careful when using automated search and replace since eV2Joule also exists in the SI namespace.

BrianMarre replied:
also needs to be renamed in the other functions

psychocoderHPC replied:
offline discussed. Other functions are named equally. I am open that someone refactor it in a sperate PR because this is out of the scope of this PR.

BrianMarre commented on include/picongpu/plugins/particleCalorimeter/ParticleCalorimeter.x.cpp:
```diff
@@ -119,8 +119,8 @@ namespace picongpu
                 const float_64 maxEnergy_SI = this->maxEnergy * sim.unit.energy();
-                const float_64 minEnergy_keV = minEnergy_SI * sim.si.conv.joule2ev(1.0e-3);
-                const float_64 maxEnergy_keV = maxEnergy_SI * sim.si.conv.joule2ev(1.0e-3);
+                const float_64 minEnergy_keV = sim.pic.conv().eV2Joule(minEnergy_SI * 1.0e-3);
+                const float_64 maxEnergy_keV = sim.pic.conv().eV2Joule(maxEnergy_SI * 1.0e-3);
```
bug
```suggestion
                const float_64 minEnergy_keV = 1.0e-3 * sim.pic.conv().joule2eV(minEnergy_SI);
                const float_64 maxEnergy_keV = 1.0e-3 * sim.pic.conv().joule2eV(maxEnergy_SI);
```

psychocoderHPC replied:
changed it to `sim.si.conv()`

BrianMarre commented on include/picongpu/plugins/particleCalorimeter/ParticleCalorimeter.x.cpp:
```diff
@@ -301,8 +301,8 @@ namespace picongpu
             /* convert units */
-            const float_64 minEnergy_SI = this->minEnergy * sim.si.conv.ev2Joule(1.0e3);
-            const float_64 maxEnergy_SI = this->maxEnergy * sim.si.conv.ev2Joule(1.0e3);
+            const float_64 minEnergy_SI = this->minEnergy * sim.si.conv().eV2Joule(1.0e3);
+            const float_64 maxEnergy_SI = this->maxEnergy * sim.si.conv().eV2Joule(1.0e3);
```
bug
```suggestion
            const float_64 minEnergy_SI = this->minEnergy * sim.unit.energy();
            const float_64 maxEnergy_SI = this->maxEnergy * sim.unit.energy();
```

psychocoderHPC replied:
it conversion to keV, I will add this to the comment

psychocoderHPC replied:
changed to 

```
            /* convert keV to joule */
            const float_64 minEnergy_SI = sim.si.conv().eV2Joule(this->minEnergy * 1.0e3);
            const float_64 maxEnergy_SI = sim.si.conv().eV2Joule(this->maxEnergy * 1.0e3);
            /* convert into PIConGPU units */
            this->minEnergy = minEnergy_SI / sim.unit.energy();
            this->maxEnergy = maxEnergy_SI / sim.unit.energy();
```

psychocoderHPC committed:
review changes second round

BrianMarre reviewed (approved):
Looks good

BrianMarre reviewed (changes requested):
Nope, the FLYonPIC rate calculation unit test fails.

For reference, check by doing the following
```bash
pic-create $PIC_EXAMPLES/../tests/compileAtomicPhysics ./testUnitSystem
git clone git@github.com:ComputationalRadiationPhysics/TestUnitSystemAtomicData.git
cd ./testUnitSystem
cp ../TestUnitSystemAtomicData/* ./
pic-build
./bin/picongpu -g 1 1 1 -d 1 1 1 -s 1 --periodic 1 1 1
```

psychocoderHPC committed:
fix wrong rydberg energy

In #5131 we changed rydberg energy from eV to joule but missed to apply
the conversion in the code.

psychocoderHPC commented:
fix is applied
